### PR TITLE
Transform to add an agent's name into the message content

### PR DESCRIPTION
**Pull request by marklysze**
The recent addition of non-OpenAI client classes has expanded the utility of AutoGen to a lot of different models through other cloud providers as well as the ability to run it with local LLMs.

However, the `name` field on each message is removed as only OpenAI can accept and utilise that field.

This means that functionality, like the GroupChat's select speaker, isn't as effective as the LLM may not be able to determine which 'agent' belongs to each message.

This PR introduces a transform that can add an agent's name to the start or end of message's text content. As this is a transform, it isn't permanent and doesn't change the original message.

How it works:
```
from autogen.agentchat.contrib.capabilities import transform_messages, transforms

# Create a transform
name_transform = TextMessageContentName(position="start", format_string="'{name}' said:\n")

# Create the TransformMessages
context_handling = transform_messages.TransformMessages(
            transforms=[
                name_transform
            ]
        )

# Add it to an agent so when they run inference it will apply to the messages
context_handling.add_to_agent(my_agent)

# Example
# Messages before
[
        {"role": "system", "content": "I am the system."},
        {"role": "user", "name": "charlie", "content": "I think the sky is blue."},
        {"role": "user", "name": "mary", "content": "The sky is red."},
        {"role": "user", "name": "bob", "content": "The sky is crimson."},
    ]

# Messages going into LLM (where name is removed, non-OpenAI clients)
[
        {"role": "system", "content": "I am the system."},
        {"role": "user", "content": "'charlie' said:\nI think the sky is blue."},
        {"role": "user", "content": "'mary' said:\nThe sky is red."},
        {"role": "user", "content": "'bob' said:\nThe sky is crimson."},
    ]
```

**Documentation to be done.**

## Dependency

For the GroupChat select speaker to be able to utilise this, PR #2719 needs to be merged first.

## Why are these changes needed?

To provide the necessary name context for non-OpenAI inference.

## Checks

- [X] I've included any doc changes needed for https://microsoft.github.io/autogen/. See https://microsoft.github.io/autogen/docs/Contribute#documentation to build and test documentation locally.
- [X] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [X] I've made sure all auto checks have passed.

(3334)

### CODE REVIEW

**Comment by sonichi** [comment]
Thanks. Could you include the code snippet in "how it works" in docstr and a test?

File: autogen/agentchat/contrib/capabilities/transforms.py [hunk 1]
```diff
@@ -445,3 +445,95 @@ def _compress_text(self, text: str) -> Tuple[str, int]:
...
+            format_string (str): The f-string to format the message name with. Use '{name}' as a placeholder for the agent's name. Defaults to '{name}:\n' and must contain '{name}'.
+            deduplicate (bool): Whether to deduplicate the formatted string so it doesn't appear twice (sometimes the LLM will add it to new messages itself). Defaults to True.
+            filter_dict (None or dict): A dictionary to filter out messages that you want/don't want to compress.
+                If None, no filters will be applied.
+            exclude_filter (bool): If exclude filter is True (the default value), messages that match the filter will be
+                excluded from compression. If False, messages that match the filter will be compressed.
+        """
+
+        assert isinstance(position, str) and position is not None
+        assert position in ["start", "end"]
```
Comment (Hk669) [inline autogen/agentchat/contrib/capabilities/transforms.py]: can we just combine these checks.

```suggestion
        assert isinstance(position, str) and position in ["start", "end"]
```

File: autogen/agentchat/contrib/capabilities/transforms.py [hunk 2]
```diff
@@ -445,3 +445,95 @@ def _compress_text(self, text: str) -> Tuple[str, int]:
...
+        assert isinstance(format_string, str) and format_string is not None
+        assert "{name}" in format_string
```
Comment (Hk669) [inline autogen/agentchat/contrib/capabilities/transforms.py]:
```suggestion
        assert isinstance(format_string, str) and "{name}" in format_string 
```

**Comment by Hk669** [comment]
@marklysze , ive also noticed, gpt-4o-mini doesnt support `name` and breaks the flow.

**Comment by marklysze** [comment]
> @marklysze , ive also noticed, gpt-4o-mini doesnt support `name` and breaks the flow.

This was noted in another issue, do you have a code sample? I couldn't replicate an issue with name included.

File: autogen/agentchat/contrib/capabilities/transforms.py (hunk 1 above)
Comment (marklysze) [reply autogen/agentchat/contrib/capabilities/transforms.py]: I'll just check that `None` is handled correctly.

Comment (marklysze) [reply autogen/agentchat/contrib/capabilities/transforms.py]: Okay, `None` handled okay in both cases, thanks - will change

**Comment by marklysze** [comment]
> Thanks. Could you include the code snippet in "how it works" in docstr and a test?

Thanks @sonichi, I've added the code snipped to the class docstring.

For the tests, the test_transforms.py has a number of tests for including the name related to position and filtering. Was there some you thought were missing? [Reference](https://github.com/autogen-ai/autogen/blob/7c98bc11d958be3cab76faaf44f2d56b0adf78b5/test/agentchat/contrib/capabilities/test_transforms.py#L64)

**Comment by marklysze** [comment]
A note: use of this transform is included in website doc changes in PR #2

**Review by sonichi — APPROVED** [review]

**Pull request merged**